envoy.dependency.check/envoy/dependency/check/abstract/issues.py

```diff
@@ -0,0 +1,270 @@
+        """Close old associated issue."""
+        # TODO: reassign any users old -> new issue
+        newer_release = await dep.newer_release
+        await old_issue.comment(
+            self.closing_tpl.format(
+                newer_release=newer_release,
+                newer_release_date=await newer_release.date,
+                full_name=dep.repo.name,
+                repo_location=self.repo_name,
+                number=self.number))
+        await old_issue.close()
+
+    async def comment(self, comment: str) -> Any:
+        """Comment on this issue."""
+        return await self.issue.comment(comment)
+
+
+class AGithubDependencyIssues(metaclass=abstracts.Abstraction):
+    """Github issues associated with dependencies."""
+
+    def __init__(
+            self,
+            github,
+            body_tpl: str = BODY_TPL,
+            closing_tpl: str = CLOSING_TPL,
+            issues_search_tpl: str = ISSUES_SEARCH_TPL,
+            labels: Tuple[str, ...] = LABELS,
+            repo_name: str = GITHUB_REPO_LOCATION,
+            title_prefix: str = TITLE_PREFIX,
+            title_re_tpl: str = TITLE_RE_TPL,
+            title_tpl: str = TITLE_TPL) -> None:
+        self.github = github
+        self.body_tpl = body_tpl
+        self.closing_tpl = closing_tpl
+        self.issues_search_tpl = issues_search_tpl
+        self.labels = labels
+        self.repo_name = repo_name
+        self.title_prefix = title_prefix
+        self.title_re_tpl = title_re_tpl
+        self.title_tpl = title_tpl
+
+    async def __aiter__(
+            self) -> AsyncGenerator[
+                AGithubDependencyIssue,
+                github.GithubIssue]:
+        async for issue in self.iter_issues():
+            issue = self.issue_class(self, issue)
+            if issue.dep:
+                yield issue
+
+    @async_property(cache=True)
+    async def dep_issues(self) -> Dict[str, AGithubDependencyIssue]:
+        """Dependency dictionary of current issues."""
+        issues: Dict[str, AGithubDependencyIssue] = {}
+        for issue in await self.open_issues:
+            if issue.dep in issues:
+                if issue.version <= issues[issue.dep].version:
+                    continue
+            issues[issue.dep] = issue
+        return issues
+
+    @async_property
+    async def duplicate_issues(
+            self) -> AsyncGenerator[
+                AGithubDependencyIssue,
+                AGithubDependencyIssue]:
+        """Iterate duplicate issues."""
+        dep_issues = await self.dep_issues
+        for issue in await self.open_issues:
+            if issue not in dep_issues.values():
+                yield issue
+
+    @property  # type:ignore
+    @abstracts.interfacemethod
+    def issue_class(self) -> Type[AGithubDependencyIssue]:
+        """Issue class."""
+        raise NotImplementedError
+
+    @async_property(cache=True)
+    async def missing_labels(self) -> Tuple[str, ...]:
```

Comment:
There is a lot of logic in this file that I think might also apply to other situations where we want to automagically create, dedupe and label issues. Would it make sense to factor this out?

Reply:
yep, that makes sense - i have made some progress on this already making it more generic so we can cve_issues

the bit im not entirely clear about is where to factor it out to - perhaps to aio.api.github if the logic is generic enough

can i do this as/when i add the cve issues as this interface will require a bit of rethinking then anyway

Reply:
Sure, just wanted to flag this as worth considering.